lte/gateway/c/session_manager/LocalEnforcer.cpp

```diff
@@ -147,13 +148,127 @@ void LocalEnforcer::setup(
+
+  // clear all the reporting flags
+  // TODO this could be done in one go with the SessionStore update below
+  session_store_.set_and_save_reporting_flag(false, request, session_uc);
+  auto updates_by_session = UpdateRequestsBySession(request);
+  if (!status.ok()) {
+    MLOG(MERROR) << "UpdateSession request to FeG/PolicyDB failed entirely: "
+                 << status.error_message();
+    handle_update_failure(*session_map_ptr, updates_by_session, session_uc);
+    report_session_update_event_failure(
+        *session_map_ptr, updates_by_session, status.error_message());
+    session_store_.update_sessions(session_uc);
+    return;
+  }
+  // Success!
+  update_session_credits_and_rules(*session_map_ptr, response, session_uc);
+  report_session_update_event(*session_map_ptr, updates_by_session);
+  session_store_.update_sessions(session_uc);
+}
+
+void LocalEnforcer::check_usage_for_reporting(
+    SessionMap session_map, SessionUpdate& session_uc) {
+  std::vector<std::unique_ptr<ServiceAction>> actions;
+  auto request = collect_updates(session_map, actions, session_uc);
+  execute_actions(session_map, actions, session_uc);
+  if (request.updates_size() == 0 && request.usage_monitors_size() == 0) {
+    auto update_success = session_store_.update_sessions(session_uc);
+    if (update_success) {
+      MLOG(MDEBUG) << "Succeeded in updating session after no reporting";
+    } else {
+      MLOG(MERROR) << "Failed in updating session after no reporting";
+    }
+    return;  // nothing to report
+  }
+
+  MLOG(MINFO) << "Sending " << request.updates_size()
+              << " charging updates and " << request.usage_monitors_size()
+              << " monitor updates to OCS and PCRF";
+
+  // set reporting flag for those sessions reporting
+  session_store_.set_and_save_reporting_flag(true, request, session_uc);
+
+  // Before reporting and returning control to the event loop, increment the
+  // request numbers stored for the sessions in SessionStore
+  session_store_.sync_request_numbers(session_uc);
+
+  reporter_->report_updates(
+      request,
+      [this, request,
+       session_map_ptr = std::make_shared<SessionMap>(std::move(session_map)),
+       session_uc](Status status, UpdateSessionResponse response) mutable {
+        report_updates_callback(
+            request, session_map_ptr, session_uc, status, response);
+      });
+}
+
+void LocalEnforcer::handle_pipelined_response(
+    Status status, RuleRecordTable resp) {
+  if (!status.ok()) {
+    MLOG(MERROR) << "Could not successfully poll stats: "
+                 << status.error_message();
+  } else {
+    auto session_map = session_store_.read_all_sessions();
+    SessionUpdate update =
+        SessionStore::get_default_session_update(session_map);
+    MLOG(MDEBUG) << "Aggregating " << resp.records_size() << " records";
+    aggregate_records(session_map, resp, update);
+
+    check_usage_for_reporting(std::move(session_map), update);
```

Comment:
Instead of `std::move` ing the entire map, you can pass by reference here.

```
void LocalEnforcer::check_usage_for_reporting(
     SessionMap& session_map, SessionUpdate& updates)
```

Reply:
Tried this before but it errors out.